pkg/virtual/workspaces/cmd/cmd.go

```diff
@@ -0,0 +1,150 @@
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	kcpinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	"github.com/kcp-dev/kcp/pkg/virtual/generic/builders"
+	virtualgenericcmd "github.com/kcp-dev/kcp/pkg/virtual/generic/cmd"
+	virtualgenericrbac "github.com/kcp-dev/kcp/pkg/virtual/generic/rbac"
+	virtualrootapiserver "github.com/kcp-dev/kcp/pkg/virtual/generic/rootapiserver"
+	virtualworkspacesbuilders "github.com/kcp-dev/kcp/pkg/virtual/workspaces/builders"
+)
+
+var _ virtualgenericcmd.SubCommandOptions = (*WorkspacesSubCommandOptions)(nil)
+
+type WorkspacesSubCommandOptions struct {
+	RootPathPrefix string
+	KubeconfigFile string
+}
+
+func (o *WorkspacesSubCommandOptions) Description() virtualgenericcmd.SubCommandDescription {
+	return virtualgenericcmd.SubCommandDescription{
+		Name:  "workspaces",
+		Use:   "workspaces",
+		Short: "Launch workspaces virtual workspace apiserver",
+		Long:  "Start a virtual workspace apiserver to managing personal, organizational or global workspaces",
+	}
+}
+
+func (o *WorkspacesSubCommandOptions) AddFlags(flags *pflag.FlagSet) {
+	if o == nil {
+		return
+	}
+
+	flags.StringVar(&o.KubeconfigFile, "workspaces:kubeconfig", "", ""+
+		"The kubeconfig file of the organizational cluster that provides the workspaces and related RBAC rules.")
+
+	_ = cobra.MarkFlagRequired(flags, "kubeconfig")
+
+	flags.StringVar(&o.RootPathPrefix, "workspaces:root-path-prefix", virtualworkspacesbuilders.DefaultRootPathPrefix, ""+
+		"The prefix of the workspaces API server root path.\n"+
+		"The final workspaces API root path will be of the form:\n    <root-path-prefix>/personal|organization|global")
+}
+
+func (o *WorkspacesSubCommandOptions) Validate() []error {
+	if o == nil {
+		return nil
+	}
+	errs := []error{}
+
+	if len(o.KubeconfigFile) == 0 {
+		errs = append(errs, errors.New("--workspaces:kubeconfig is required for this command"))
+	}
+
+	if !strings.HasPrefix(o.RootPathPrefix, "/") {
+		errs = append(errs, fmt.Errorf("--workspaces:root-path-prefix %v should start with /", o.RootPathPrefix))
+	}
+
+	return errs
+}
+
+func (o *WorkspacesSubCommandOptions) InitializeBuilders() ([]virtualrootapiserver.InformerStart, []builders.VirtualWorkspaceBuilder, error) {
+	kubeConfig, err := virtualgenericcmd.ReadKubeConfig(o.KubeconfigFile)
+	if err != nil {
+		return nil, nil, err
+	}
+	kubeClientConfig, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	apiExtensionsClient, err := apiextensionsclient.NewForConfig(kubeClientConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	workspacesCRD, err := apiExtensionsClient.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), "workspaces."+tenancyAPI.SchemeGroupVersion.Group, metav1.GetOptions{})
+	if kerrors.IsNotFound(err) {
+		return nil, nil, errors.New("The Workspaces CRD should be registered in the cluster providing the workspaces")
+	}
```

Comment:
this should rather use discovery, not checking for a CRD.

Reply:
I sort-of agree here, but in any case this is a temporary workaround. And it also allows providing the logical cluster name of the organisation workspace we will search the workspaces into, since it is currently not possible to guess it (since it includes the shard hostname prefix is not a predictible name).

Reply:
So we'll have to change this part of code in a next PR when things are ready for it.